worker/controller/manifold.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package controller
+
+import (
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	worker "gopkg.in/juju/worker.v1"
+	"gopkg.in/tomb.v1"
+
+	coreagent "github.com/juju/juju/agent"
+	"github.com/juju/juju/state"
+	"github.com/juju/juju/worker/dependency"
+)
+
+var logger = loggo.GetLogger("juju.worker.controller")
+
+// ManifoldConfig provides the dependencies for Manifold.
+type ManifoldConfig struct {
+	AgentName              string
+	StateConfigWatcherName string
+	OpenController         func(coreagent.Config) (*state.Controller, error)
+	PingInterval           time.Duration
+}
+
+const defaultPingInterval = 15 * time.Second
+
+// Manifold returns a manifold whose worker which wraps a
+// *state.State, which is in turn wrapper by a Tracker.  It will
+// exit if the State's associated mongodb session dies.
+func Manifold(config ManifoldConfig) dependency.Manifold {
+	return dependency.Manifold{
+		Inputs: []string{
+			config.AgentName,
+			config.StateConfigWatcherName,
+		},
+		Start: func(context dependency.Context) (worker.Worker, error) {
+			// First, a sanity check.
+			if config.OpenController == nil {
+				return nil, errors.New("OpenController is nil in config")
+			}
+
+			// Get the agent.
+			var agent coreagent.Agent
+			if err := context.Get(config.AgentName, &agent); err != nil {
+				return nil, err
+			}
+
+			// Confirm we're running in a state server by asking the
+			// stateconfigwatcher manifold.
+			var haveStateConfig bool
+			if err := context.Get(config.StateConfigWatcherName, &haveStateConfig); err != nil {
+				return nil, err
+			}
+			if !haveStateConfig {
+				return nil, dependency.ErrMissing
+			}
+
+			ctlr, err := config.OpenController(agent.CurrentConfig())
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			tracker := newTracker(ctlr)
+
+			pingInterval := config.PingInterval
```

Comment:
This feels like it should be slightly more contextual "MongoPingInterval", unless its something just about the config of this exact worker.
But 'what are you pinging' is definitely unclear.